### PR TITLE
Closes #1435: Add support for alert popup for GeckoEngine...

### LINKED ISSUE #1435: Add support for alert popup for GeckoEngine
User Story: https://github.com/mozilla-mobile/reference-browser/issues/223

### PR DESCRIPTION
**Pull request by Amejia481**
Depends on: https://github.com/mozilla-mobile/android-components/pull/1437

### CODE REVIEW

**Comment by Amejia481** [comment]
[Video Sample](https://drive.google.com/file/d/1H9fYug5b9KTw-RNx7jR1J5WWhHv8T_ql/view?usp=sharing)

**Comment by pocmo** [comment]
@Amejia481 This is now a thing that WebView doesn't implement by default. Can you look at `WebChromeClient.onJsAlert()` and see if we could map that to the same generic prompt API? We do not necessarily need to implement that now but we should make sure that we do not end up with something that will make it super hard to do later. :)

**Comment by Amejia481** [comment]
Sure, let me take a look!

**Comment by Amejia481** [comment]
@pocmo Yes, we can use the same API, the only thing is that we have to implement our own logic for handling when a session should be prevented for showing more alert, because webview doesn't provide one and gecko does.

[WebView quick impl](https://drive.google.com/file/d/1dLljIV5413ZHNMngOxPOGHCaYZwpNfAi/view?usp=sharing)

**Comment by Amejia481** [comment]
Issue for WebView work: https://github.com/mozilla-mobile/android-components/issues/1504

**Comment by pocmo** [comment]
Sounds good. Thanks 👍

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt [hunk 1]
```diff
@@ -78,6 +79,26 @@ internal class GeckoPromptDelegate(private val geckoEngineSession: GeckoEngineSe
         }
     }
 
+    override fun onAlert(
+        session: GeckoSession?,
+        title: String?,
+        message: String?,
+        callback: AlertCallback?
```
Comment (csadilek) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt]: It doesn't look like callback can actually be null. You will get some coverage improvements too if you make it non nullable.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AlertDialogFragment.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    companion object {
+        /**
+         * A builder method for creating a [AlertDialogFragment]
+         * @property sessionId to create the dialog.
+         * @property title the title of the dialog.
+         * @property message the message of the dialog.
+         * @property manyDialogs tells if this [sessionId] has shown many dialogs is short period of time,
+         * if is true a checkbox will be part of the dialog, for the user to choose
+         * if wants to prevent this [sessionId] continuing showing dialogs.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AlertDialogFragment.kt]: Hmm this seems like an unneeded indirection. GeckoView tells us whether or not to display a Checkbox, which we turn into `manyDialogs`, just to document that it's used to decide whether or not to display a checkbox.

Maybe just use rename to `showMoreDialogsCheckbox`?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt [hunk 3]
```diff
@@ -29,4 +29,20 @@ sealed class PromptRequest {
...
      */
     data class MenuChoice(val choices: Array<Choice>, val onSelect: (Choice) -> Unit) : PromptRequest()
+
+    /**
+     * Value type that represents a request for a alert prompt.
+     * @property title of the dialog.
+     * @property message the body of the dialog.
+     * @property manyDialogs tells if this page has shown multiple prompts within a short period of time.
+     * @property onDismiss callback to let the page know the user dismiss the dialog.
+     * @property onShouldShowNoMoreDialogs tells the web page if should continue showing alerts or not.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt]: `if it...`

File: components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AlertDialogFragmentTest.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        val messageTextView = dialog.findViewById<TextView>(android.R.id.message)
+        val checkBox = dialog.findViewById<CheckBox>(id.no_more_dialogs_check_box)
+
+        assertEquals(titleTextView.text, "title")
+        assertEquals(messageTextView.text, "message")
+        assertNotNull(checkBox)
+    }
+
+    @Test
+    fun `Alert with manyManyDialogs equals false should not have a checkbox`() {
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/prompts/src/test/java/mozilla/components/feature/prompts/AlertDialogFragmentTest.kt]: 🙃

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AlertDialogFragment.kt (hunk 2 above)
Comment (Amejia481) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/AlertDialogFragment.kt]: What about hasShownManyDialogs or ShouldShowCheckBox?

**Comment by Amejia481** [comment]
@csadilek  I updated the PR :)

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 5]
```diff
@@ -65,8 +66,8 @@ class PromptFeature(
      * Event that is triggered when a native dialog needs to be shown.
      * Displays suitable dialog for the type of the [promptRequest].
      *
-     * @property session The session the request the dialog.
-     * @property promptRequest The session the request the dialog.
+     * @param session The session the request the dialog.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: The session which requested the dialog?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 6]
```diff
@@ -94,8 +101,8 @@ class PromptFeature(
 
     /**
      * Event that is triggered when a single choice or menu item is selected in a dialog.
-     * @property sessionId the id for which session was selected.
-     * @property choice the selection from the dialog.
+     * @param sessionId the id for which session was selected.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: This is the id of the session which requested the prompt?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 7]
```diff
@@ -111,8 +118,8 @@ class PromptFeature(
 
     /**
      * Event that is triggered when a multiple choice items are selected in a dialog.
-     * @property sessionId the id for which session was selected.
-     * @property choices the selected items from the dialog.
+     * @param sessionId the id for which session was selected.
+     * @param choices the selected items from the dialog.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: This is the id of the session which requested the prompt?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt [hunk 8]
```diff
@@ -126,11 +133,32 @@ class PromptFeature(
     /**
      * Event that is called when a dialog is dismissed.
      * This consumes the [PromptFeature] value from the [Session] indicated by [sessionId].
-     * @property sessionId the id for which session was dismissed.
+     * @param sessionId the id for which session was dismissed.
```
Comment (csadilek) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt]: This is the id of the session which requested the prompt?

**Comment by Amejia481** [comment]
No worries the idea is to improve :)

**Pull request merged**